### ISSUE TITLE
Trying to build the OF 3DPrimitivesExample using CMake.

### ISSUE THREAD

**Issue by symbolix** [kureta/oF-CLion#1]
Hi Kureta,

I have been following your CMake project for a while and I am trying to come up with a CMake workflow that will allow me to build OF projects using pure CMake and no other IDE. I have adapted one of your templates that I have cloned from @judepereira.

As a starting point, I am trying to build the `3DPrimitivesExample` provided with OF. However, I have not succeeded so far. The example builds and runs fine through XCode. So I am attempting to use your approach and also fill in the gaps with digging into Xcode to see how things are linked etc.

So far, I managed to build OpenFrameworks.a using XCode, so all the libs and headers are there. However, when I try to build the example using a CMake only approach, I am getting lots of warnings and linker errors etc. However, the most severe linker error is the one where the compiler complains about wrong architecture (missing symbols) etc. Below is the CMakeLists.txt file that I have started to work on based on your CMakeLists.txt.

``` cmake
cmake_minimum_required(VERSION 2.8.12)
project(3DPrimitiveExample)
set(APP_NAME 3DPrimitiveExample)
set(BUNDLE_NAME "3DPrimitiveExample")

# ===================== oF Directory =====================
set(OF_DIRECTORY "/Users/me/packages/builds/x86_64/of-0.9.3-osx-release")

# ===================== Compiler Flags =====================
set(CMAKE_C_COMPILER "/usr/bin/clang")
set(CMAKE_C_FLAGS "")

set(CMAKE_CXX_COMPILER "/usr/bin/clang++")
set(CMAKE_CXX_FLAGS "-stdlib=libstdc++ -D__MACOSX_CORE__")

set(CMAKE_OSX_ARCHITECTURES x86_64)
add_compile_options(-Wno-deprecated-declarations)

# ===================== Source Files =====================
set(SOURCE_FILES
        src/main.cpp
        src/ofApp.cpp)

# =========== Set Global Properties ==============
MESSAGE("++ [debug]: OF_DIRECTORY: ${OF_DIRECTORY}")

# ===================== System Frameworks =====================

# ===================== Include Dirs =====================
# ;--------------------------------------------------------------------------------
# ; External Headers
# ;--------------------------------------------------------------------------------
set(HEADER_FREEIMAGE ${OF_DIRECTORY}/libs/FreeImage/include)
set(HEADER_BOOST ${OF_DIRECTORY}/libs/boost/include)
set(HEADER_CAIRO ${OF_DIRECTORY}/libs/cairo/include/cairo)
set(HEADER_FMODEX ${OF_DIRECTORY}/libs/fmodex/include)
set(HEADER_FREETYPE ${OF_DIRECTORY}/libs/freetype/include)
set(HEADER_GLEW ${OF_DIRECTORY}/libs/glew/include)
set(HEADER_GLFW ${OF_DIRECTORY}/libs/glfw/include)
set(HEADER_GLUT ${OF_DIRECTORY}/libs/glut/include)
set(HEADER_SSL ${OF_DIRECTORY}/libs/openssl/include)
set(HEADER_POCO ${OF_DIRECTORY}/libs/poco/include)
set(HEADER_RTAUDIO ${OF_DIRECTORY}/libs/rtAudio/include)
set(HEADER_TESS2 ${OF_DIRECTORY}/libs/tess2/include)
set(HEADER_UTF8 ${OF_DIRECTORY}/libs/utf8cpp/include)

# ;--------------------------------------------------------------------------------
# ; OpenFrameworks Native Headers
# ;--------------------------------------------------------------------------------
set(HEADER_OF_0 ${OF_DIRECTORY}/libs/openFrameworks)
set(HEADER_OF_1 ${OF_DIRECTORY}/libs/openFrameworks/3d)
set(HEADER_OF_2 ${OF_DIRECTORY}/libs/openFrameworks/app)
set(HEADER_OF_3 ${OF_DIRECTORY}/libs/openFrameworks/communication)
set(HEADER_OF_4 ${OF_DIRECTORY}/libs/openFrameworks/events)
set(HEADER_OF_5 ${OF_DIRECTORY}/libs/openFrameworks/gl)
set(HEADER_OF_6 ${OF_DIRECTORY}/libs/openFrameworks/graphics)
set(HEADER_OF_7 ${OF_DIRECTORY}/libs/openFrameworks/math)
set(HEADER_OF_8 ${OF_DIRECTORY}/libs/openFrameworks/sound)
set(HEADER_OF_9 ${OF_DIRECTORY}/libs/openFrameworks/types)
set(HEADER_OF_10 ${OF_DIRECTORY}/libs/openFrameworks/utils)
set(HEADER_OF_11 ${OF_DIRECTORY}/libs/openFrameworks/video)

# Concatenate variable.
set(OF_CORE_HEADERS
    ${HEADER_OF_0}
    ${HEADER_OF_1}
    ${HEADER_OF_2}
    ${HEADER_OF_3}
    ${HEADER_OF_4}
    ${HEADER_OF_5}
    ${HEADER_OF_6}
    ${HEADER_OF_7}
    ${HEADER_OF_8}
    ${HEADER_OF_9}
    ${HEADER_OF_10}
    ${HEADER_OF_11}
    ${HEADER_FREEIMAGE}
    ${HEADER_BOOST}
    ${HEADER_CAIRO}
    ${HEADER_FMODEX}
    ${HEADER_FREETYPE}
    ${HEADER_GLEW}
    ${HEADER_GLFW}
    ${HEADER_GLUT}
    ${HEADER_SSL}
    ${HEADER_POCO}
    ${HEADER_RTAUDIO}
    ${HEADER_TESS2}
    ${HEADER_UTF8})

# ===================== Libs =====================
# ;--------------------------------------------------------------------------------
# ; OpenFrameworks Native Headers
# ;--------------------------------------------------------------------------------
set(LIB_OF ${OF_DIRECTORY}/libs/openFrameworksCompiled/lib/osx/openFrameworks.a)
# ;--------------------------------------------------------------------------------
# ; FreeImage
# ;--------------------------------------------------------------------------------
set(LIB_FREEIMAGE ${OF_DIRECTORY}/libs/FreeImage/lib/osx/freeimage.a)
# ;--------------------------------------------------------------------------------
# ; Boost
# ;--------------------------------------------------------------------------------
set(LIB_BOOST_MAIN ${OF_DIRECTORY}/libs/boost/lib/osx/boost.a)
set(LIB_BOOST_FILESYSTEM ${OF_DIRECTORY}/libs/boost/lib/osx/boost_filesystem.a)
set(LIB_BOOST_SYSTEM ${OF_DIRECTORY}/libs/boost/lib/osx/boost_system.a)
# ;--------------------------------------------------------------------------------
# ; Cairo
# ;--------------------------------------------------------------------------------
set(LIB_CAIRO1 ${OF_DIRECTORY}/libs/cairo/lib/osx/cairo-script-interpreter.a)
set(LIB_CAIRO2 ${OF_DIRECTORY}/libs/cairo/lib/osx/cairo.a)
set(LIB_CAIRO3 ${OF_DIRECTORY}/libs/cairo/lib/osx/pixman-1.a)
set(LIB_CAIRO4 ${OF_DIRECTORY}/libs/cairo/lib/osx/png.a)
# ;--------------------------------------------------------------------------------
# ; Fmodex
# ;--------------------------------------------------------------------------------
set(LIB_FMODEX ${OF_DIRECTORY}/libs/fmodex/lib/osx/libfmodex.dylib)
# ;--------------------------------------------------------------------------------
# ; FreeType
# ;--------------------------------------------------------------------------------
set(LIB_FREETYPE ${OF_DIRECTORY}/libs/freetype/lib/osx/freetype.a)
# ;--------------------------------------------------------------------------------
# ; Glew, Glfw, Glut
# ;--------------------------------------------------------------------------------
set(LIB_GLEW ${OF_DIRECTORY}/libs/glew/lib/osx/glew.a)
set(LIB_GLFW ${OF_DIRECTORY}/libs/glfw/lib/osx/glfw3.a)
# No libs for GLUT.
# ;--------------------------------------------------------------------------------
# ; OpenSSL
# ;--------------------------------------------------------------------------------
set(LIB_OPENSSL1 ${OF_DIRECTORY}/libs/openssl/lib/osx/crypto.a)
set(LIB_OPENSSL2 ${OF_DIRECTORY}/libs/openssl/lib/osx/ssl.a)
# ;--------------------------------------------------------------------------------
# ; Poco
# ;--------------------------------------------------------------------------------
set(LIB_POCO_1 ${OF_DIRECTORY}/libs/poco/lib/osx/PocoCrypto.a)
set(LIB_POCO_2 ${OF_DIRECTORY}/libs/poco/lib/osx/PocoData.a)
set(LIB_POCO_3 ${OF_DIRECTORY}/libs/poco/lib/osx/PocoDataSQLite.a)
set(LIB_POCO_4 ${OF_DIRECTORY}/libs/poco/lib/osx/PocoFoundation.a)
set(LIB_POCO_5 ${OF_DIRECTORY}/libs/poco/lib/osx/PocoJSON.a)
set(LIB_POCO_6 ${OF_DIRECTORY}/libs/poco/lib/osx/PocoMongoDB.a)
set(LIB_POCO_7 ${OF_DIRECTORY}/libs/poco/lib/osx/PocoNet.a)
set(LIB_POCO_8 ${OF_DIRECTORY}/libs/poco/lib/osx/PocoNetSSL.a)
set(LIB_POCO_9 ${OF_DIRECTORY}/libs/poco/lib/osx/PocoUtil.a)
set(LIB_POCO_10 ${OF_DIRECTORY}/libs/poco/lib/osx/PocoXML.a)
set(LIB_POCO_11 ${OF_DIRECTORY}/libs/poco/lib/osx/PocoZip.a)
# ;--------------------------------------------------------------------------------
# ; RtAudio
# ;--------------------------------------------------------------------------------
set(LIB_RTAUDIO ${OF_DIRECTORY}/libs/rtAudio/lib/osx/rtAudio.a)
# ;--------------------------------------------------------------------------------
# ; Tess2
# ;--------------------------------------------------------------------------------
set(LIB_TESS ${OF_DIRECTORY}/libs/tess2/lib/osx/tess2.a)
# ;--------------------------------------------------------------------------------
# ; utf8cpp
# ;--------------------------------------------------------------------------------
# No libs for utf8cpp.

# Concatenate variable.
set(OF_CORE_LIBS
    ${LIB_OF}
    ${LIB_FREEIMAGE}
    ${LIB_BOOST_MAIN}
    ${LIB_BOOST_FILESYSTEM}
    ${LIB_BOOST_SYSTEM}
    ${LIB_CAIRO1}
    ${LIB_CAIRO2}
    ${LIB_CAIRO3}
    ${LIB_CAIRO4}
    ${LIB_FMODEX}
    ${LIB_FREETYPE}
    ${LIB_GLEW}
    ${LIB_GLFW} 
    ${LIB_OPENSSL1}
    ${LIB_OPENSSL2}
    ${LIB_POCO_1}
    ${LIB_POCO_2}
    ${LIB_POCO_3}
    ${LIB_POCO_4}
    ${LIB_POCO_5}
    ${LIB_POCO_6}
    ${LIB_POCO_7}
    ${LIB_POCO_8}
    ${LIB_POCO_9}
    ${LIB_POCO_10}
    ${LIB_POCO_11}
    ${LIB_RTAUDIO}
    ${LIB_TESS})

# ====================== Prep Linking ==========================
include_directories (${OF_CORE_HEADERS})
link_directories (${OF_CORE_LIBS})

set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wextra -std=c++11 -stdlib=libc++ -O2 -arch x86_64 -Wall")

# ============ Compile and Link ==================
add_executable(${APP_NAME} ${SOURCE_FILES})
target_link_libraries(${APP_NAME} ${OF_CORE_LIBS})

# ============ MACOSX_BUNDLE Settings ============
set_target_properties(${APP_NAME} PROPERTIES MACOSX_BUNDLE_INFO_STRING "Cmake openFrameworks CLion test app")
set_target_properties(${APP_NAME} PROPERTIES MACOSX_BUNDLE_LONG_VERSION_STRING "0.0.1 pre-release alpha")
set_target_properties(${APP_NAME} PROPERTIES MACOSX_BUNDLE_BUNDLE_NAME ${BUNDLE_NAME})
set_target_properties(${APP_NAME} PROPERTIES MACOSX_BUNDLE_SHORT_VERSION_STRING 0.0.1)
set_target_properties(${APP_NAME} PROPERTIES MACOSX_BUNDLE_BUNDLE_VERSION 0.0.1)
set_target_properties(${APP_NAME} PROPERTIES MACOSX_BUNDLE_COPYRIGHT "(c) kureta 2015")
set_target_properties(${APP_NAME} PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${PROJECT_SOURCE_DIR}/bin)
```

I am sure I have broken something, so I will go back and start from scratch.

Thanks.

**Comment by symbolix** [comment]
When I reverted back to your original CMakeLists.txt designed to compile an OF Example Project, I am getting the following error:

```
CMake Error: The source directory "/Users/me/packages/builds/x86_64/of-0.9.3-osx-release" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
make[2]: *** [of_shared-prefix/src/of_shared-stamp/of_shared-configure] Error 1
make[1]: *** [CMakeFiles/of_shared.dir/all] Error 2
make: *** [all] Error 2
```

It seems like you are linking to the external project, which contains the main **openFrameworks** folders. However, I have already compiled openFrameworks using Xcode, and all the libs are ready. At what stage I should be including the libs for the `link_directories (${OF_CORE_LIBS})` and `target_link_libraries()`. All my libs, including the openFrameworks one, are static. Except the `libfmodex`, which is dynamic.

Thanks.

**Comment by judepereira** [comment]
@symbolix: The path to the oF release dir is incorrect:
/Users/**me**/**packages/builds/x86_64/of-0.9.3-osx-release**

Download an oF release, and then update that path

**Comment by symbolix** [comment]
Hello @judepereira 
Thank you very much for your time. This is where my oF lives. I have already downloaded it and built it using XCode. So now all the headers are here:

```
; oF Headers:
/Users/me/packages/builds/x86_64/of-0.9.3-osx-release/libs/openFrameworks
/Users/me/packages/builds/x86_64/of-0.9.3-osx-release/libs/openFrameworks/3d)
/Users/me/packages/builds/x86_64/of-0.9.3-osx-release/libs/openFrameworks/app)
/Users/me/packages/builds/x86_64/of-0.9.3-osx-release/libs/openFrameworks/communication)
/Users/me/packages/builds/x86_64/of-0.9.3-osx-release/libs/openFrameworks/events)
/Users/me/packages/builds/x86_64/of-0.9.3-osx-release/libs/openFrameworks/gl)
/Users/me/packages/builds/x86_64/of-0.9.3-osx-release/libs/openFrameworks/graphics)
/Users/me/packages/builds/x86_64/of-0.9.3-osx-release/libs/openFrameworks/math)
/Users/me/packages/builds/x86_64/of-0.9.3-osx-release/libs/openFrameworks/sound)
/Users/me/packages/builds/x86_64/of-0.9.3-osx-release/libs/openFrameworks/types)
/Users/me/packages/builds/x86_64/of-0.9.3-osx-release/libs/openFrameworks/utils)
/Users/me/packages/builds/x86_64/of-0.9.3-osx-release/libs/openFrameworks/video)

; Third Party Headers: 
/Users/me/packages/builds/x86_64/of-0.9.3-osx-release/libs/${lib}/include
```

And the oF static lib is here:

```
/Users/me/packages/builds/x86_64/of-0.9.3-osx-release/libs/openFrameworksCompiled/lib/osx/openFrameworks.a
```

And all the other static / dynamic libs are here:

```
 /Users/me/packages/builds/x86_64/of-0.9.3-osx-release/libs/boost/lib/osx/boost_filesystem.a
 /Users/me/packages/builds/x86_64/of-0.9.3-osx-release/libs/FreeImage/lib/osx/freeimage.a
(etc.)
```

So when I run the example from inside Xcode, it builds and compiles fine and runs fine. However, I am only using Xcode just to make sure things are working.

The only thing I can think of is put the `oF-CLion/openFrameworks/CMakeLists.txt` file inside the oF root folder and let the example CMakeLists.txt include it as a shared component and rebuild all of the openFrameworks as a dynamic lib. I can try that? ... or maybe I misunderstood you?

Right now, I am trying to get away with only one single `CMakeLists.txt` that will just build the example without rebuilding the entire oF.

Once again thank you very much for the help.

**Comment by judepereira** [comment]
@symbolix Oh okay. Sorry, I thought that was a placeholder in the path.
Can you add the CMakeLists.txt from here to your oF project and try?
https://github.com/kureta/oF-CLion/blob/master/openFrameworks/CMakeLists.txt

**Comment by symbolix** [comment]
Hi @judepereira 
I will try now. I will put together a tiny repo with my CMake file. I am really desperate and I would like to get this task done, so thank you very much for the help. Just to clarify, I am trying to build the **3DPrimitivesExample** using CMake.

Yesterday, I got quite a lot of progress. Most of the linker errors are gone now, so the includes and the static libs are working, however I am getting symbol errors on one of the libs, glfw one. I will post an update.

**Comment by symbolix** [comment]
So, that CMakeLists.txt is for building oF, right? I have a CMakeLists.txt to build my example. How exactly should I combine those? Do you mean, that I should put the one you have mentioned inside the oF folder? Cause, I suppose, the CMakeList that builds the oF example will try to call the one that builds oF, right?

**Comment by judepereira** [comment]
Yes, we've used it as an external CMake project. Hence, you've got to have a CMakeLists.txt in the oF root directory too.

**Comment by symbolix** [comment]
Ok. Thanks for the advice. I will try that. The part that I do not understand is, why I would need to have the `CMakeLists.txt` inside the **oF root** folder if I already have **oF** built as a static lib.

So I have the `openFrameworks.a` library file and this is the `lib` side. I also have the headers etc. and this is the `include` bit. For everything else (the external dependencies), like `boost`, `pico`, `glfw` etc, I have the `.a` files, which are the static libs, then I have the header files for those as well, which provide the `include` side of the solution. In theory, passing those to the compiler through a `CMakeLists.txt` that is designed to compile the `3DPrimitivesExample` should be sufficient. (... and just to make it clear, I am talking about the `libs`and `includes` shipped with `openFrameworks`, not the the standalone versions.)

I have the feeling that the solution would be a combination of bits and pieces from your `CMakeLists.txt`. In order to test couple of things I have recently downloaded the latest `glfw` lib and built that using CMake. Then I have written a `FindGLFW.cmake` just to test the idea of finding `libs` and `includes` and learn a bit of CMake. As a result, I have successfully built and linked and example GLFW application on OSX. It means I am pretty close. Once I find the solution, I will share it. Meanwhile, any advice is extremely welcome. Thank you!

**Comment by symbolix** [comment]
Hello Again,

I had the chance to try again. So I have the **application** CMakeLists.txt in the example application folder and the **openFrameworks** CMakeLists.txt in my Open Frameworks folder. And the **addon** CMake files are in the addons.

Things seem to be working better but I am getting the following error:

```
make[2]: *** [of_shared-prefix/src/of_shared-stamp/of_shared-configure] Error 1 
make[1]: *** [CMakeFiles/of_shared.dir/all] Error 2
make: *** [all] Error 2
```

Any ideas about what might have gone wrong? Thanks.

**Comment by judepereira** [comment]
@symbolix Paste the logs before that. What you've posted is not enough for me to take a guess at.

**Comment by symbolix** [comment]
Sure. I will try again this evening and will provide the information here as soon as possible. Thanks.

**Comment by symbolix** [comment]
Hi @judepereira 
Some good news:

First I decided to test if the `oF-CLion/openFrameworks/CMakeLists.txt` will build the SHARED library `libopenFrameworks.dylib`. On my first try, the process failed with `FreeImage` and `cairo` reporting a wrong architecture (i386 was failing). Then I commented the following line `set(CMAKE_OSX_ARCHITECTURES i386)` out and the build process was complete. I got one warning about the deprecated `QTKit` framework, but this was expected.

The only bit that is a bit broken is the bit below as the `make install` command failed to find those destinations:

```
install(FILES $<TARGET_FILE:of_shared> DESTINATION oF/lib)
install(FILES libs/fmodex/lib/osx/libfmodex.dylib DESTINATION oF/lib)
install(DIRECTORY libs/glut/lib/osx/GLUT.framework DESTINATION .)
```

However, `libopenFrameworks.dylib` was successfully created inside `${PROJECT_SOURCE_DIR/bin}`. Now the only thing I need to understand is how `of_shared` is imported inside the `CMakeLists.txt` that is supposed to build the `ofApplication`. Thanks for the support so far.

**Comment by symbolix** [comment]
Hello Again,
Update: So now I kind of understand the link between the `ofApplication/CMakeLists.txt` and the `openFrameworks/CMakeLists.txt`. The first defines the install prefix etc. and imports the second one as an external project. The build and linking seems to be working now, however the `install` stage fails:

```
[100%] Built target core
Scanning dependencies of target of_shared
Linking CXX shared library /Users/muhittinbilginer/packages/builds/x86_64/of-0.9.3-osx-release/bin/libopenFrameworks.dylib
ld: warning: ignoring file /System/Library/Frameworks//QuickTime.framework/QuickTime, file was built for i386 which is not the architecture being linked (x86_64): /System/Library/Frameworks//QuickTime.framework/QuickTime
[100%] Built target of_shared
[ 70%] Performing install step for 'of_shared'
[100%] Built target core
[100%] Built target of_shared
Install the project...
-- Install configuration: ""
 CMake Error at cmake_install.cmake:31 (FILE):
 file INSTALL cannot find
 "/Users/me/packages/builds/x86_64/of-0.9.3-osx-release/$<TARGET_FILE:of_shared>".

make[3]: *** [install] Error 1
make[2]: *** [of_shared-prefix/src/of_shared-stamp/of_shared-install] Error 2
make[1]: *** [CMakeFiles/of_shared.dir/all] Error 2
make: *** [all] Error 2
```

I can feel it, I am really close, just need to tweak the `install` stage so that everything fall in place.
:)

**Comment by judepereira** [comment]
Hey @symbolix, so you're right on track!

However, one key thing that you've missed is - oF cannot do a 64 bit build on OSX, as it uses the deprecated QTKit. When I used to work on oF projects in Xcode and Appcode, I remember that I had to set the active architectures to i386. Only then would the build work.

**Comment by symbolix** [comment]
It finally worked! Although there are still minor issues, I finally managed to build **openFrameworks** as a shared lib and build the example app purely using CMake.

I think they changed things a bit, when I check the architecture of all libs included, they all list as 64bit and 32bit architectures. So everything builds as 64bit now. The only 32bit part is QTKit and it is deprecated. I am on OF9.3.0 by the way.

The only issues are:
1) The target file install for the generated shared lib does not work. My question on stackoverflow is [here](http://stackoverflow.com/questions/38192939/target-file-fails-during-install-stage-with-cmake)

2) RLinking stage to the shared libs once everything is built is probably broken, I had to re-link the libs manually.

3) Any data folders for textures etc. are not defined in the CMake process, so those need to be copied into the bin folder manually. I will add that bit.

Apart from that, the 3DPrimitivesExample builds and runs without issues, enterely using CMake! Thanks.

**Comment by symbolix** [comment]
I have also fixed couple of more aspects to this and I have recently successfully built an `openFrameworks` app against a static `openFrameworks` lib. Now both STATIC and SHARED approaches work. Also I am working on separating the main `openFrameworks` build process and the `app` build process. It is not very effective to build **oF** every time an app is built. I understand that _building_ and _linking_ only takes place after big changes and _making_ is relatively quick, I still believe that they should be separate stages.
